mne/stats/cluster_level.py

```diff
@@ -1880,23 +1885,18 @@ def cluster_test(
             raise ValueError("for paired t-test, each subject must have 2 observations")
 
     # extract the data from the dataframe
-    def _extract_data_array(series):
+    outer_func = np.concatenate if is_epo else np.array
+    axes = (-3, -1) if is_tfr else (-2, -1)
+
+    def func_arr(series):
         return np.concatenate(series.values)
 
-    def _extract_data_mne(series):  # 2D data
-        return np.array(
-            series.map(lambda inst: inst.get_data().swapaxes(-2, -1)).to_list()
+    def func_mne(series):
+        return outer_func(
+            series.map(lambda inst: inst.get_data().swapaxes(*axes)).to_list()
         )
 
-    def _extract_data_tfr(series):
-        return series.map(lambda inst: inst.get_data().swapaxes(-3, -1)).to_list()
-
-    if _dtype is np.ndarray:
-        func = _extract_data_array
-    elif _dtype is BaseTFR:
-        func = _extract_data_tfr
-    else:
-        func = _extract_data_mne
```

Comment:
[REFACTORED] I realized I was going to need to write a fourth extraction function (in addition to `_extract_data_array`, `_extract_data_tfr`, and `_extract_data_mne`) to handle Epochs and EpochsTFRs. Looking at the `mne` and `tfr` ones we'd already written, I noticed the TFR one was wrong (needed to be wrapped in an `np.array`), which made it clear how similar those two were ---  only different in the `swapaxes` arguments --- and the new func for Epochs was only going to differ in the outer wrapper function (`np.concatenate` instead of `np.array`).

So now we can have two variables: `outer_func` and `axes` and generate the extraction function we need based on whether `is_epo` and `is_tfr`.

Reply:
I think the test confirms that we are in fact doing the right thing:

```py
n_epo, n_chan, n_freq, n_times = 6, 3, 4, 5
```

since each dimension has a different size, I think that if we'd done something wrong then the test would fail.